File: app/src/main/java/com/example/ba2023/model/CountDownModel.java
package com.example.ba2023.model;

import android.content.Context;
import android.content.Intent;
import android.os.CountDownTimer;
import android.widget.TextView;

import com.example.ba2023.FinishedActivity;
import com.example.ba2023.MainActivity;
import com.example.ba2023.ScreenActivity;

import java.util.concurrent.TimeUnit;

public class CountDownModel  extends CountDownTimer {

    private static String hms;
    private static CountDownModel instance;
    private static TextView currentTextView;
    private static Context context;

    private CountDownModel(long millisInFuture, long countDownInterval) {
        super(millisInFuture, countDownInterval);
        hms = formatTime(millisInFuture);
    }

    public static CountDownModel initInstance(long millisInFuture, long countDownInterval, Context activity) {
        if (instance == null) {
            instance = new CountDownModel(millisInFuture, countDownInterval);
            context = activity;
        }
        return instance;
    }

    public static CountDownModel getInstance() throws Exception {
        if (instance == null) {
            throw new Exception("Parameters not initialized. Initiate with initInstance");
        } else {
            return instance;
        }
    }
    public static boolean isInstanceNull() {
        return instance == null;
    }
    public static String getFormatedTime() {
        return hms;
    }
    public static String getTimeMS() {
        return hms.substring(3);
    }

    public static TextView getCurrentTextView() {
        return currentTextView;
    }

    public static void setCurrentTextView(TextView currentTextView) {
        CountDownModel.currentTextView = currentTextView;
    }

    private String formatTime(long millis) {
        String format = String.format("%02d:%02d:%02d", TimeUnit.MILLISECONDS.toHours(millis),
                TimeUnit.MILLISECONDS.toMinutes(millis) - TimeUnit.HOURS.toMinutes(TimeUnit.MILLISECONDS.toHours(millis)),
                TimeUnit.MILLISECONDS.toSeconds(millis) - TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(millis)));
        return format;
    }
    @Override
    public void onTick(long l) {

        long millis = l;
        hms = formatTime(millis);
        if (currentTextView != null) {
            currentTextView.setText(getTimeMS());
        }
    }

    @Override
    public void onFinish() {
        Intent intent = new Intent(context, FinishedActivity.class);
        context.startActivity(intent);
    }
}

File: app/src/main/java/com/example/ba2023/model/CountDownModel.kt
package com.example.ba2023.model

import WritingStatusManager
import android.content.Context
import android.content.Intent
import android.os.CountDownTimer
import android.os.VibrationEffect
import android.os.Vibrator
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.example.ba2023.FinishedActivity
import com.example.ba2023.PauseActivity
import com.example.ba2023.WritingActivity
import com.example.ba2023.model.CycleUtil.getCycle
import com.example.ba2023.model.CycleUtil.setCycle
import java.lang.ref.WeakReference
import java.util.concurrent.TimeUnit

class CountDownModel private constructor(
    millisInFuture: Long,
    countDownInterval: Long,
    context: Context
) : CountDownTimer(millisInFuture, countDownInterval) {
    private val contextRef: WeakReference<Context>

    init {
        formatedTime = formatTime(millisInFuture)
        contextRef = WeakReference(context)
    }

    private fun formatTime(millis: Long): String {
        return String.format(
            "%02d:%02d:%02d", TimeUnit.MILLISECONDS.toHours(millis),
            TimeUnit.MILLISECONDS.toMinutes(millis) - TimeUnit.HOURS.toMinutes(
                TimeUnit.MILLISECONDS.toHours(
                    millis
                )
            ),
            TimeUnit.MILLISECONDS.toSeconds(millis) - TimeUnit.MINUTES.toSeconds(
                TimeUnit.MILLISECONDS.toMinutes(
                    millis
                )
            )
        )
    }

    override fun onTick(l: Long) {
        formatedTime = formatTime(l)
        if (currentTextView != null) {
            currentTextView!!.text = timeMS
        }
    }

    override fun onFinish() {
        var cycle = getCycle()
        var switchActivity: Class<*> = PauseActivity::class.java
        //write last screen state to file
        WritingStatusManager.updateTimesSpentInStates(System.currentTimeMillis())
        //stop the exercise
        if (cycle == 0) {
            switchActivity = FinishedActivity::class.java
        }
        val context = contextRef.get()
        if (caller == PauseActivity::class.java.name) {
            WritingStatusManager.setScreenState(WritingStatusManager.ScreenState.WRITING)
            switchActivity = WritingActivity::class.java
            setCycle(--cycle)
            vibrate(context,500)

        }
        if (context != null) {
            val intent = Intent(context, switchActivity)
            context.startActivity(intent)
        }
    }


    companion object {
        lateinit var formatedTime: String
        private var instance: CountDownModel? = null
        var currentTextView: TextView? = null
        var caller: String? = null
        fun initInstance(
            millisInFuture: Long,
            countDownInterval: Long,
            activity: Context
        ): CountDownModel? {
            instance = CountDownModel(millisInFuture, countDownInterval, activity)
            return instance
        }

        @Throws(Exception::class)
        fun getInstance(): CountDownModel? {
            return if (instance == null) {
                throw Exception("Parameters not initialized. Initiate with initInstance")
            } else {
                instance
            }
        }

        val isInstanceNull: Boolean
            get() = instance == null
        val timeMS: String
            get() = formatedTime.substring(3)
    }
    private fun vibrate(context: Context?,durationMillis: Long) {
        if (context != null) {
            val vibrator = ContextCompat.getSystemService(context, Vibrator::class.java) as Vibrator
            val vibrationEffect = VibrationEffect.createOneShot(durationMillis, VibrationEffect.DEFAULT_AMPLITUDE)
            vibrator.vibrate(vibrationEffect)
        }

    }

}